src/test/java/integration/IntegrationTest.java

```diff
@@ -3,563 +3,1428 @@
+        // Set some invalid auth info; avoid null checks.
+        ws3.setAuthInfo("", "");
+    }
+
+    @Test
+    public void integrationTest() throws Exception {
+        // Test valid flow
+        testUserRegister();
+        testUserRegisterDuplicate();
+        testUserLogin(ws1, user1ID, user1Pass);
+        testUserLogin(ws2, user2ID, user2Pass);
+        testUserLoginInvalidCredentials();
+        testProjectGetPermissionConstants();
+        testProjectGetPermissionConstantsInvalid();
+        testUserLookup();
+        testUserLookupInvalid();
+        testUserProjects();
+        testUserProjectsInvalid();
+        testProjectCreate();
+        testProjectCreateInvalid();
+        testUserProjects();
+        testProjectLookup(ws1, user1ID, proj1);
+        testProjectLookup(ws2, user2ID, proj2);
+        testProjectGetFiles(); // should have 0 files
+        testProjectGetFilesInvalid();
+        testProjectSubscribe(ws1, proj1);
+        testProjectSubscribe(ws2, proj2);
+        testProjectSubscribeInvalid();
+        testProjectGrantPermission(ws1, proj1, user2ID, "read", new WSManager[]{ws2});
+        testProjectCrossSubscribeInvalid();
+        testProjectSubscribe(ws2, proj1);
+        testProjectCrossSubscribeInvalid(); // Check to make sure User1 cannot subscribe to Proj2
+        testProjectRename();
+        testProjectRenameInvalid();
+
+        testFileCreate(ws1, file1, proj1, fileData1, new WSManager[]{ws2});
+        testProjectGetFiles(); // should have 1 file in Proj1, 0 in Proj2
+        testFileCreate(ws2, file2, proj2, fileData2, new WSManager[]{});
+        testProjectGetFiles(); // should have 1 file in Proj1 and Proj2
+        testFileCreateInvalid();
+        testFilePull(); // Should have no changes
+        testFilePullInvalid();
+        testFileChange(ws1, file1, new WSManager[]{ws2}); // Test User1 changing File1
+        testFilePull(); // Should have 1 change
+        testFileChange(ws2, file2, new WSManager[]{}); // Test User2 changing File1
+        testFilePull(); // Should have 1 changes on both files
+        testFileChangeInvalid();
+        testProjectGrantPermission(ws1, proj1, user2ID, "write", new WSManager[]{ws2});
+        testFileChange(ws2, file1, new WSManager[]{ws1}); // Test User2 changing File1
+        testFilePull(); // Should have 2 changes on File1, 1 change on File2
+
+        // Test File Moves
+        testFileMove(ws1, file1, "newPathWS1", new WSManager[]{ws2});
+        testProjectGetFiles();
+        testFileMove(ws2, file1, "newPathWS2", new WSManager[]{ws1});
+        testProjectGetFiles();
+        testFileMove(ws2, file2, "newPathWS2", new WSManager[]{});
+        testProjectGetFiles();
+        testFileMoveInvalid();
+
+        testFileChange(ws1, file1, new WSManager[]{ws2}); // Test User1 changing File1
+        testFilePull(); // Should have 3 changes on File1, 1 change on File2
+        testFileChange(ws2, file2, new WSManager[]{}); // Test User2 changing File2
```

Comment:
Might be worth it to change method signatures that take an array of type WSManager to a varargs parameter so that you don't have to pass in an empty array (also it just looks cleaner in method calls).